test/gtest/ucp/test_ucp_proto.cc

```diff
@@ -563,4 +605,224 @@ UCS_TEST_P(test_proto_perf_add_func, intersect_middle)
+
+private:
+    // Generate a random ascending sequence of numbers, ending with SIZE_MAX
+    static std::vector<size_t> generate_random_sequence(unsigned length);
+
+    // Generate a random perf structure
+    static perf_ptr_t generate_random_perf(unsigned max_segments);
+
+    // Collect points of sampling based on perf structure segments
+    static std::vector<size_t> get_sample_points(perf_ptr_t perf);
+
+    // Calculate the expected aggregation result at a given point. If the
+    // expected result is undefined, return nullptr.
+    std::unique_ptr<perf_factors_map_t>
+    expected_aggregate_result(const std::vector<perf_ptr_t> &perfs,
+                              size_t point);
+};
+
+std::vector<size_t>
+test_proto_perf_aggregate::generate_random_sequence(unsigned length)
+{
+    std::vector<size_t> sequence;
+    size_t value = 0;
+    std::generate_n(std::back_inserter(sequence), length, [&value]() {
+        value += ucs::rand_range(10000);
+        return value;
+    });
+    sequence.emplace_back(SIZE_MAX);
+    return sequence;
+}
+
+test_proto_perf::perf_ptr_t
+test_proto_perf_aggregate::generate_random_perf(unsigned max_segments)
+{
+    perf_ptr_t perf = create();
+    size_t start    = 0;
+    auto points     = generate_random_sequence(max_segments);
+    for (auto point : points) {
+        if (ucs::rand_range(4) > 0) {
+            add_func(perf, start, point, UCP_PROTO_PERF_FACTOR_LOCAL_TL,
+                     local_tl_func);
+        }
+        start = point + 1;
+    }
+    return perf;
+}
+
+std::vector<size_t>
+test_proto_perf_aggregate::get_sample_points(perf_ptr_t perf)
+{
+    std::vector<size_t> points;
+
+    auto seg = find_lb(perf, 0);
+    while (seg != NULL) {
+        auto start = ucp_proto_perf_segment_start(seg);
+        auto end   = ucp_proto_perf_segment_end(seg);
+        points.push_back(start);
+        points.push_back(end);
+        points.push_back((start + end) / 2);
+        if (start > 0) {
+            points.push_back(start - 1);
+        }
+        if (end == SIZE_MAX) {
+            break;
+        }
+
+        points.push_back(end + 1);
+        seg = find_lb(perf, end + 1);
+    }
+
+    return points;
+}
+
+std::unique_ptr<test_proto_perf::perf_factors_map_t>
+test_proto_perf_aggregate::expected_aggregate_result(const perf_vec_t &perfs,
+                                                     size_t point)
+{
+    perf_factors_map_t agg;
+    for (auto perf : perfs) {
+        ucp_proto_perf_segment_t *seg = find_lb(perf, point);
+        if ((seg == NULL) || (point < ucp_proto_perf_segment_start(seg))) {
+            return nullptr; // Point not found in one of the perf objects
+        }
+
+        for (int factor_id = 0; factor_id < UCP_PROTO_PERF_FACTOR_LAST;
+             ++factor_id) {
+            auto func = ucp_proto_perf_segment_func(
+                    seg, (ucp_proto_perf_factor_id_t)factor_id);
+            if (!ucs_linear_func_is_zero(func, UCP_PROTO_PERF_EPSILON)) {
```

Comment:
Can we remove this check?

Reply:
it's used to simplify the output in case of getting invalid result